vm/systemSmartContracts/staking.go

```diff
@@ -1592,41 +1649,81 @@ func (s *stakingSC) stakeNodesFromQueue(args *vmcommon.ContractCallInput) vmcomm
 
 	s.addToStakedNodes(int64(stakedNodes))
 
+	err = s.cleanAdditionalQueueNotEnoughFunds(waitingListData, mapCheckedOwners)
+	if err != nil {
+		s.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
 	return vmcommon.Ok
 }
 
+type validatorFundInfo struct {
+	numNodesToUnstake uint32
+	qualifiedToStake  bool
+}
+
 func (s *stakingSC) checkValidatorFunds(
-	mapCheckedOwners map[string]bool,
+	mapCheckedOwners map[string]*validatorFundInfo,
 	owner []byte,
-) (bool, error) {
-	hasFunds, okInMap := mapCheckedOwners[string(owner)]
+) (*validatorFundInfo, error) {
+	validatorInfo, okInMap := mapCheckedOwners[string(owner)]
 	if okInMap {
-		return hasFunds, nil
+		return validatorInfo, nil
 	}
 
 	marshaledData := s.eei.GetStorageFromAddress(s.stakeAccessAddr, owner)
 	if len(marshaledData) == 0 {
-		mapCheckedOwners[string(owner)] = false
-		return false, nil
+		validatorInfo = &validatorFundInfo{
+			numNodesToUnstake: math.MaxUint32,
+			qualifiedToStake:  false,
+		}
+		mapCheckedOwners[string(owner)] = validatorInfo
+		return validatorInfo, nil
 	}
 
 	validatorData := &ValidatorDataV2{}
 	err := s.marshalizer.Unmarshal(validatorData, marshaledData)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
 
 	numRegisteredKeys := int64(len(validatorData.BlsPubKeys))
 	numQualified := big.NewInt(0).Div(validatorData.TotalStakeValue, s.minNodePrice).Int64()
 
-	if numQualified < numRegisteredKeys {
-		mapCheckedOwners[string(owner)] = false
-		return false, nil
+	if numQualified >= numRegisteredKeys {
+		validatorInfo = &validatorFundInfo{
+			numNodesToUnstake: 0,
+			qualifiedToStake:  true,
+		}
+		mapCheckedOwners[string(owner)] = validatorInfo
+		return validatorInfo, nil
 	}
 
-	mapCheckedOwners[string(owner)] = true
+	numActiveNodes := int64(0)
+	for _, blsKey := range validatorData.BlsPubKeys {
+		stakedData, errGet := s.getOrCreateRegisteredData(blsKey)
+		if errGet != nil {
+			return nil, errGet
+		}
+
+		if stakedData.Staked || stakedData.Waiting || stakedData.Jailed {
```

Comment:
why do we consider stakedData.Waiting as an active node?
is this referring to staking queue or shard waiting list?
if it is referring to staking queue then I think this is incorrect.

Reply:
Waiting is considered a node for which 2500eGLD was paid. It is referring to the staking queue and this function is correct. The same is used when a validator stakes a node - it is verified if there is enough funds and all of its nodes are qualified and whether you can add one more or not.

Reply:
so an active node here is a key with at least 2500 staked, even though that key is not currently part of the 3200 allowed active nodes?

Then the naming is confusing, can we rename this to numRegisteredNodes instead of numActiveNodes?

Reply:
done